include/allocator.h

```diff
@@ -0,0 +1,67 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license.
+
+#pragma once
+#include <atomic>
+#include <memory>
+#include <vector>
+#include <unordered_map>
+#include "tsl/robin_set.h"
+#include "windows_customizations.h"
+
+namespace diskann
+{
+
+/// <summary>
+/// The wrapper of std::allocator which tracks the memory usage
+/// </summary>
+template <class T> class Allocator
+{
+  public:
+    using value_type = T;
+    using propagate_on_container_move_assignment = std::true_type;
+    using is_always_equal = std::true_type;
+
+    Allocator(std::shared_ptr<std::atomic<size_t>> &memory_used_in_bytes_ptr = nullptr)
+    {
+        _memory_used_in_bytes_ptr = memory_used_in_bytes_ptr;
+    }
+
+    template <class U> Allocator(const Allocator<U> &a)
+    {
+        _memory_used_in_bytes_ptr = a._memory_used_in_bytes_ptr;
+    }
+
+    DISKANN_DLLEXPORT T *allocate(std::size_t count, const T *hint = nullptr);
+
+    DISKANN_DLLEXPORT void deallocate(T *ptr, std::size_t count);
+
+    std::shared_ptr<std::atomic<size_t>> _memory_used_in_bytes_ptr;
+
+  private:
+    void update_memory_usage(std::size_t count, bool is_allocation);
+
+    std::allocator<T> _allocator;
+};
+
+template <class T> inline bool operator==(const Allocator<T> &a, const Allocator<T> &b)
```

Comment:
These operations should probably also call the underlying operator==/operator!= on _allocator